client/go/outline/routingtable/ip_routing_table.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2025 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package routingtable
+
+import (
+	"container/list"
+	"fmt"
+	"net/netip"
+)
+
+// Compile-time check
+var _ RoutingTable[netip.Prefix, netip.Addr, any] = (*IPRoutingTable[any])(nil)
+
+// "V" is typically expected to be a dialer of some kind
+type IPRoutingTable[V any] struct {
+	prefixBuckets map[int]map[netip.Prefix]V
+	prefixLengths *list.List
```

Comment:
We need separate lists for IPv4 and IPv6, or else one may override the other if they have the same prefix.

Also, I think it's simpler if you just drop the prefixLength list. It's a small list, you can iterate over all 32-128 values quickly. In that case the top-level map becomes an array.